interpreter/cling/lib/Interpreter/Value.cpp

```diff
@@ -255,11 +265,46 @@ namespace cling {
     const clang::ASTContext& ctx = getASTContext();
     unsigned payloadSize = ctx.getTypeSizeInChars(getType()).getQuantity();
     m_Storage.m_Ptr = AllocatedValue::CreatePayload(payloadSize, dtorFunc,
-                                                    GetNumberOfElements());
+                                                GetNumberOfElements(getType()));
+  }
+
+  void Value::AssertTypeMismatch(const char* Type) const {
+#ifndef NDEBUG
+    assert(isBuiltinType() && "Must be a builtin!");
+    const clang::BuiltinType *BT = getType()->castAs<clang::BuiltinType>();
+    clang::PrintingPolicy Policy = getASTContext().getPrintingPolicy();
+#endif // NDEBUG
+    assert(BT->getName(Policy).equals(Type));
   }
 
+  static clang::QualType getCorrespondingBuiltin(clang::ASTContext &C,
+                                                 clang::BuiltinType::Kind K) {
+    using namespace clang;
+    switch(K) {
+    default:
+      assert(false && "Type not supported");
+      return {};
+#define BUILTIN_TYPE(Id, SingletonId) \
+      case BuiltinType::Id: return C.SingletonId;
+#include "clang/AST/BuiltinTypes.def"
+    }
+  }
+
+#define X(type, name)                                                   \
+  template <> Value Value::Create(Interpreter& Interp, type val) {      \
```

Comment:
See your discussion with Richard a while back, on not inlining template specializations (in the context of rootcling's dictionary explicit function template  specializations). Again a good motivation to just use overloads for these, rather than template specializatins.

Reply:
I tried to find it back for a couple of minutes, in vain. @pcanal might remember - this was about putting the implementation for a specialization into the source file, with a prior declaration... But fine, if you are convinced that you want this to be a template then let's keep it this way!